src/cogs/CourseThreads.py

```diff
@@ -0,0 +1,268 @@
+    async def leave_course(
+        self, ctx: commands.Context, course: Union[Course, None] = None
+    ):
+        """
+        Leave a course thread, if you're in it. Note that this is idempotent.
+        """
+        if not await self._does_course_exist(ctx, course):
+            return
+        course_thread_id: int = self.course_mappings[course.year_level][
+            CURRENT_COURSES_KEY
+        ][str(course)]
+        course_thread: discord.Thread = self.client.get_channel(course_thread_id)
+        await course_thread.remove_user(ctx.author)
+        return await ctx.reply(
+            f"Done! Removed you from {course_thread.mention}, whether you were in it or not."
+        )
+
+    @courses.command(name="list", aliases=["l"])
+    @commands.guild_only()
+    async def list_courses(self, ctx: commands.Context):
+        """List all the courses that currently have a course thread."""
+        course_listing: List[str] = []
+        for year, year_metadata in self.course_mappings.items():
+            course_listing.append(f"**Level `{year}xx`**")
+            for course, channel_id in year_metadata[CURRENT_COURSES_KEY].items():
+                channel: discord.Thread = self.client.get_channel(channel_id)
+                if not channel:
+                    course_listing.append(
+                        f"  - `{course}`: {channel_id} (error getting thread)"
+                    )
+                else:
+                    course_listing.append(f"  - `{course}`: {channel.mention}")
+        await Paginator(title="Available Courses", entries=course_listing).paginate(ctx)
+
+    @courses.command(name="search", aliases=["s"])
+    @commands.guild_only()
+    async def search_courses(self, ctx: commands.Context, query: str):
+        """Searches the thread directory for a case-insensitive match."""
+        search_results: List[str] = []
+        for year_metadata in self.course_mappings.values():
+            for course, channel_id in year_metadata[CURRENT_COURSES_KEY].items():
+                if (
+                    query.lower() in course.lower()
+                    or query.lower() in course.replace(" ", "").lower()
+                ):
+                    channel: discord.Thread = self.client.get_channel(channel_id)
+                    if not channel:
+                        search_results.append(
+                            f"- `{course}`: {channel_id} (error getting thread)"
+                        )
+                    else:
+                        search_results.append(f"- `{course}`: {channel.mention}")
+        if not search_results:
+            return await ctx.reply(
+                f"No courses found for `{query}`.",
+                allowed_mentions=discord.AllowedMentions.none(),
+            )
+        await Paginator(
+            title=f"Courses Matching: `{query}`", entries=search_results
+        ).paginate(ctx)
+
+    @tasks.loop(seconds=1)
+    async def thread_refresher_task(self):
```

Comment:
What's the consequence if a thread is unarchived and a `!course list` is called? I assume this shouldn't happen given we're constantly iterating through all the threads but just curious.

Reply:
It wouldn't be part of the list of threads this task will watch, so it will act like a regular Discord thread -- that is, it will archive after `n` time period of inactivity.